src/StringOrArrayDynamicFunctionReturnTypeExtension.php

```diff
@@ -29,21 +35,63 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
     // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter
     public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
     {
-        $args = $functionCall->getArgs();
-        if (count($args) === 0) {
+        if (count($functionCall->getArgs()) === 0) {
             return null;
         }
 
-        $dataArgType = $scope->getType($args[0]->value);
-        if ($dataArgType->isArray()->yes()) {
-            $keyType = $dataArgType->getIterableKeyType();
-            if ($keyType instanceof StringType) {
-                return new ArrayType(new StringType(), new StringType());
+        return TypeTraverser::map(
+            $scope->getType($functionCall->getArgs()[0]->value),
+            function (Type $type, callable $traverse): Type {
+                if ($type instanceof UnionType || $type instanceof IntersectionType) {
+                    // phpcs:ignore NeutronStandard.Functions.VariableFunctions.VariableFunction
+                    return $traverse($type);
+                }
+
+                return $this->getType($type);
             }
+        );
+    }
+
+    private function getType(Type $type): Type
+    {
+        if ($type->isScalar()->yes()) {
+            return new StringType();
+        }
+
+        if (!$type->isArray()->yes()) {
```

Comment:
```suggestion
        if ($type->isArray()->no()) {
```

Reply:
If it is a maybe type !yes means maybe or no,
!no means maybe or yes.

Reply:
Both are not a 100% correct I suppose, the maybe case might need another extra handling instead of just returning the input below. I'll check it later.

Reply:
dealing with the maybe here is highly annoying, e.g. your proposed change would lead to a `*NEVER*` (mine to a `''`) when a maybe array like `mixed` or `iterable` is passed. both are wrong. and in both cases PHPStan should be complaining because esc_sql does not support those types via PHPDoc. Natively it does of course, I know..

And I can't even say what we would return with a mixed, I guess we could just bail out early and not handle it or smth like that

Reply:
adapted, what do you think? should at least be simpler and less wrong now :)

Reply:
@IanDelMar we listen to you

Reply:
> And I can't even say what we would return with a mixed, I guess we could just bail out early and not handle and return the default type or smth like that

Yep, I would go for returning null if everything else does not apply. That's what we're doing in other extensions as well, no? Dealing with what we want to deal with and return the default otherwise. It's still an improvement and doesn't do worse in any case.

Reply:
Thought the same, but can't return null here because this is wrapped with the TypeTraverser:map() which doesn't allow returning null AFAIK

Reply:
BTW just curious, your profile says Vienna but you seem yo be active very late until into the night. Ate you still in Vienna? I'm from Austria and lived there too, missing it a bit from time to time..

Reply:
I still live in Vienna. My activity time mainly depends on my day job and after work napping :-D. Vienna is a nice place to live. I would miss it too at least in late spring and summer.

Reply:
And good old boss Viktor is in Budapest, the twin cities, right? :)
Anyway, shall we merge? :P

Reply:
I *am* old.